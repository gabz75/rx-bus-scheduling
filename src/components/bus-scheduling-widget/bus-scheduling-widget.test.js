import React from 'react';
import { render } from '@testing-library/react';

import data from 'data/bus-scheduling-input.json';

import BusWidgetScheduling from './index';

const ACTIVE_STYLE = 'background-color: #015baa';

it('renders', () => {
  const { asFragment } = render(<BusWidgetScheduling trips={data} />);
  expect(asFragment()).toMatchSnapshot();
});

it('moves the first trip to the second bus', () => {
  const { getAllByTestId } = render(<BusWidgetScheduling trips={data} />);
  expect(getAllByTestId('bus-row')).toHaveLength(9);
  expect(getAllByTestId('button-trip')).toHaveLength(9);

  const [firstTrip, secontTrip] = getAllByTestId('button-trip');

  expect(firstTrip).not.toHaveStyle(ACTIVE_STYLE); // not active
  expect(getAllByTestId('bus-row')[0]).toContainElement(firstTrip); // row1 contains trip1
  expect(getAllByTestId('bus-row')[1]).toContainElement(secontTrip); // row2 contains trip2

  getAllByTestId('button-trip')[0].click();

  expect(firstTrip).toHaveStyle(ACTIVE_STYLE); // trip1 becomes active

  getAllByTestId('bus-row')[1].click();

  expect(getAllByTestId('bus-row')).toHaveLength(8); // we lost a row
  expect(getAllByTestId('button-trip')).toHaveLength(9);

  const [newFirstTrip, newSecondTrip] = getAllByTestId('button-trip');
  expect(getAllByTestId('bus-row')[0]).toContainElement(newFirstTrip); // first row contains both trips
  expect(getAllByTestId('bus-row')[0]).toContainElement(newSecondTrip);
});

it('creates a new row when selecting a trip', () => {
  const { getAllByTestId } = render(<BusWidgetScheduling trips={data} />);
  expect(getAllByTestId('bus-row')).toHaveLength(9);
  expect(getAllByTestId('button-trip')).toHaveLength(9);

  getAllByTestId('button-trip')[0].click();

  expect(getAllByTestId('bus-row')).toHaveLength(10);
});

it('prevents from moving trip that overlaps', () => {
  const { getAllByTestId } = render(<BusWidgetScheduling trips={data} />);
  expect(getAllByTestId('bus-row')).toHaveLength(9);
  expect(getAllByTestId('button-trip')).toHaveLength(9);

  const firstTrip = getAllByTestId('button-trip')[1];

  expect(firstTrip).not.toHaveStyle(ACTIVE_STYLE); // not active

  // 2nd trip overlaps with 4th trip.
  firstTrip.click();

  expect(firstTrip).toHaveStyle(ACTIVE_STYLE); // still active becaus the trip conflicts

  getAllByTestId('bus-row')[3].click();

  expect(firstTrip).toHaveStyle(ACTIVE_STYLE); // still active becaus the trip conflicts
});
