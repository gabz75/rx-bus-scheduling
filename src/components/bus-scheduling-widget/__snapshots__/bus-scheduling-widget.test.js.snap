// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="sc-AykKE jZChVS"
  >
    <div
      class="sc-AykKF ljrprT"
    >
      <span
        style="left: 0px;"
      >
        0:00
      </span>
      <span
        style="left: 60px;"
      >
        1:00
      </span>
      <span
        style="left: 120px;"
      >
        2:00
      </span>
      <span
        style="left: 180px;"
      >
        3:00
      </span>
      <span
        style="left: 240px;"
      >
        4:00
      </span>
      <span
        style="left: 300px;"
      >
        5:00
      </span>
      <span
        style="left: 360px;"
      >
        6:00
      </span>
      <span
        style="left: 420px;"
      >
        7:00
      </span>
      <span
        style="left: 480px;"
      >
        9:00
      </span>
    </div>
    <div
      class="sc-AykKG bhca-DG"
    >
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 30px; width: 120px;"
          type="button"
        >
          1
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 180px; width: 120px;"
          type="button"
        >
          2
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 330px; width: 120px;"
          type="button"
        >
          3
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 40px; width: 90px;"
          type="button"
        >
          4
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 160px; width: 90px;"
          type="button"
        >
          5
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 280px; width: 90px;"
          type="button"
        >
          6
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 400px; width: 90px;"
          type="button"
        >
          7
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 80px; width: 160px;"
          type="button"
        >
          8
        </button>
      </div>
      <div
        class="sc-AykKC dmQAbO"
        data-testid="bus-row"
      >
        <button
          class="sc-AykKD bPHaww"
          data-testid="button-trip"
          style="left: 280px; width: 150px;"
          type="button"
        >
          9
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
